### PR TITLE
Bug 1890838: Condense indexmanagement into single cronjob

### PR DESCRIPTION
**Pull request by jcantrill**
### Description
This PR collapses the multiple policy cronjobs to a single job with multiple tasks it runs:
* delete
* rollover

The reasoning is there is a potential race condition between the previous jobs which both rely upon a `-write` alias that may lead to false information.  Additionally, ES does not have transactions or is ACID.  By converting these into tasks we execute for management we:
* potentially free disk for ES to do additional work
* give a better chance for the rollover to be successful
* May need to add a task to "unblock" the read-only index identified in the BZ

One risk in this proposed solution is failure of the first container will block any subsequent behaviors.  Placing the PR on hold pending agreement its a viable solution

/cc @lukas-vlcek @periklis 
/assign  @ewolinetz 
/assign @alanconway 
/cherrypick release-4.6

### Links
https://bugzilla.redhat.com/show_bug.cgi?id=1890838

/hold

### CODE REVIEW

**Review by periklis — CHANGES_REQUESTED** [review]
For the record as mentioned previously, I suggest to merge the two scripts in one script with proper error handling between the two stages `delete` and `rollover` instead of splitting these two task in different container types of a pod. The rationale to keep them in a single container is:
1. All error/info logs of the IM cronjob land on a single container log. (Reminder: init containers maintain own container log)
2. We can handle errors programatically, e.g. using bash/python now later refactoring this to a single golang program.
3. Introducing an init container while backporting will put the ask from release engineering on resources (CPU/Mem), etc. 

In summary, I believe composing the two scripts into one is the way of least resistance w/ a minimum in effort and doc update needed to explain why we IM is now a single job.

**Bonus:** This PR is missing to delete the old cronjobs before introducing the unified one.

File: internal/indexmanagement/reconcile.go
Comment (ewolinetz) [inline internal/indexmanagement/reconcile.go:74]: doesn't this mean we won't remove the old cronjobs that we no longer support?

File: internal/indexmanagement/reconcile.go
Comment (ewolinetz) [inline internal/indexmanagement/reconcile.go:140]: accidentally left in? or is github not formatting correctly...

File: internal/indexmanagement/reconcile.go
Comment (ewolinetz) [inline internal/indexmanagement/reconcile.go:151]: this seems misleading... its that we aren't configuring for delete, right?

File: internal/indexmanagement/reconcile.go
Comment (ewolinetz) [inline internal/indexmanagement/reconcile.go:166]: similar as above

File: internal/indexmanagement/reconcile.go
Comment (ewolinetz) [inline internal/indexmanagement/reconcile.go:170]: not sure this is buying us anything given we have messages above that would indicate we aren't doing hot/delete phases

Comment (ewolinetz) [reply internal/indexmanagement/reconcile.go:170]: should we instead be returning `nil`?

File: internal/indexmanagement/reconcile.go
Comment (ewolinetz) [inline internal/indexmanagement/reconcile.go:303]: why are we using init containers?

File: internal/k8shandler/index_management.go
Comment (ewolinetz) [inline internal/k8shandler/index_management.go:53]: "indexmanagement"

File: internal/indexmanagement/reconcile.go
Comment (jcantrill) [reply internal/indexmanagement/reconcile.go:74]: This code produces a set of cronjobs we expect.  It will list those deployed and then remove those which are not in the expected set.  I had the same concerns but observed the operator removing the old jobs.  It would be worth QE (@anpingli)  verifying this to be the case

File: internal/indexmanagement/reconcile.go
Comment (jcantrill) [reply internal/indexmanagement/reconcile.go:140]: no this looks to be a valid error.  will correct

File: internal/indexmanagement/reconcile.go
Comment (jcantrill) [reply internal/indexmanagement/reconcile.go:151]: updating message to try and be more clear... `log.V(1).Info("Skipping curation management for policymapping; delete phase not defined", "policymapping", mapping.Name)`

File: internal/indexmanagement/reconcile.go
Comment (jcantrill) [reply internal/indexmanagement/reconcile.go:166]: `log.V(1).Info("Skipping rollover management for policymapping; hot phase not defined", "policymapping", mapping.Name)`

File: internal/indexmanagement/reconcile.go
Comment (jcantrill) [reply internal/indexmanagement/reconcile.go:170]: This only buys us an additional message that we are not creating the job at all in lieu of only constructing a portion of the job.  I can be swayed to remove it if you so desire

File: internal/indexmanagement/reconcile.go
Comment (jcantrill) [reply internal/indexmanagement/reconcile.go:303]: Ref comment 0: `One risk in this proposed solution is failure of the first container will block any subsequent behaviors. Placing the PR on hold pending agreement its a viable solution`

Init containers run sequentially which is what we want; we want to stay out of a race condition where we are rolling over and moving the write alias but also trying to remove indices.  Regular containers would run in parallel.  The downside to this approach is the running of the second container (rollover) is always contingent upon the successful completion of delete.  I'm not convinced this is necessarily the best approach but was the quickest path to a change without modifying any scripts.  It could be we are better served accepting a "delete" failure and then just continuing to try and rollover so that both continue to happen.

File: internal/indexmanagement/reconcile.go
Comment (ewolinetz) [reply internal/indexmanagement/reconcile.go:170]: I think since we are already going to be printing out that we are skipping "delete" and "hot" for when no phases are defined we can drop it... unless we do a check before this if both "hot" and "delete" are missing we would just print this message and then return early?

also, can we adjust the message? "no phases not defined"

File: internal/indexmanagement/reconcile.go
Comment (ewolinetz) [reply internal/indexmanagement/reconcile.go:303]: i think we want to move away from the init container approach. can't we update our scripts to be blocking and not require the prior action to pass? (e.g. if delete fails we will still try to roll over)

File: internal/indexmanagement/reconcile.go
Comment (jcantrill) [reply internal/indexmanagement/reconcile.go:170]: Just noticed... i think there is a return nil in this block which was lost during rebase.

File: internal/indexmanagement/reconcile.go
Comment (jcantrill) [reply internal/indexmanagement/reconcile.go:303]: Init containers are a convenient mechanism to isolate specific tasks and I believe it to be a good application here aside from the need to possibly run task 2 regardless of task 1.  I will pose this question before making mods because I don't know the best answer:  is there a drawback to continuing to rollover if delete were to fail?

Issue i could see is we add more shards thus consuming more RAM in lieu of simply allowing the index to increase in size.  That likely has its own limitations but it may be lesser issue.  I defer to @lukas-vlcek here to provide some feedback

Comment (ewolinetz) [reply internal/indexmanagement/reconcile.go:303]: if we don't rollover we risk the index being written to becoming large and exceeding recommended sizes

File: internal/indexmanagement/reconcile.go
Comment (ewolinetz) [reply internal/indexmanagement/reconcile.go:170]: i think its above, there was a comment about it

File: internal/indexmanagement/reconcile.go
Comment (jcantrill) [reply internal/indexmanagement/reconcile.go]: per our discussion, move logic from init containers to single container running scripts

File: internal/indexmanagement/reconcile.go
Comment (ewolinetz) [inline internal/indexmanagement/reconcile.go:141]: "no phases not defined" --> "no phases are defined" ?

File: internal/indexmanagement/reconcile.go
Comment (ewolinetz) [inline internal/indexmanagement/reconcile.go:36]: surprised this wasn't auto-formatted to be consistent with the above lines

File: internal/indexmanagement/reconcile.go
Comment (ewolinetz) [inline internal/indexmanagement/reconcile.go:218]: shouldn't this be a `Log.Error` ?

File: internal/indexmanagement/reconcile.go
Comment (ewolinetz) [inline internal/indexmanagement/reconcile.go:256]: we can drop this

File: internal/indexmanagement/reconcile.go
Comment (ewolinetz) [inline internal/indexmanagement/reconcile.go:259]: same here

**Comment by ewolinetz** [comment]
@jcantrill one thing we also want to do in this is to adjust the rollover script to do a sanity check after calling the rollover api

File: internal/indexmanagement/reconcile.go
Comment (jcantrill) [reply internal/indexmanagement/reconcile.go:218]: removing.  hold over from testing

File: internal/indexmanagement/reconcile.go
Comment (jcantrill) [reply internal/indexmanagement/reconcile.go:36]: fixed

**Comment by jcantrill** [comment]
> @jcantrill one thing we also want to do in this is to adjust the rollover script to do a sanity check after calling the rollover api

IMO we should do that in a separate PR against the other BZ

**Pull request merged**
